Nodejs/Product/Nodejs/SharedProject/HierarchyIdMap.cs

```diff
@@ -57,33 +47,21 @@ public void Remove(HierarchyNode node)
         {
             VisualStudio.Shell.ThreadHelper.ThrowIfNotOnUIThread();
 
-            if (node == null)
-            {
-                throw new ArgumentNullException(nameof(node));
-            }
+            Debug.Assert(node != null, "Called with null node");
 
-            lock (this.theLock)
-            {
-                var i = (int)node.ID - 1;
-                if (0 > i || i >= this.ids.Count)
-                {
-                    throw new InvalidOperationException($"Invalid id. {i}");
-                }
+            var idx = (int)node.ID - 1;
 
-                var weakRef = this.ids[i];
-                if (weakRef == null)
-                {
-                    throw new InvalidOperationException("Trying to retrieve a node before adding.");
-                }
+            var removeCheck = this.nodes.TryGetValue(idx, out var weakRef);
 
-                if (weakRef.TryGetTarget(out var found) && !object.ReferenceEquals(node, found))
-                {
-                    throw new InvalidOperationException("The node has the wrong id.");
-                }
+            Debug.Assert(removeCheck, "How did we get an id, which we haven't seen before");
```

Comment:
>removeCheck [](start = 25, length = 11)

Since `removeCheck == (weakRef != null)`, this also indicates a double delete, right?

Reply:
weakref may be null, because we don't know have that id in the dictionary, i.e. never added, or because we set it to null previously, i.e. double delete.

Reply:
Null no longer indicates a double-delete.

---
In reply to: [182285218](https://github.com/Microsoft/nodejstools/pull/1933#discussion_r182285218) [](ancestors = 182285218)